enhancements/machine-api/aws-imds-v2-support.md

```diff
@@ -0,0 +1,260 @@
+
+### Goals
+
+- provide mechanisms to enable/enforce IMDSv2 for newly created machines via Machine API
+- provide mechanisms to enable/enforce IMDSv2 for master machines during installation procedure for newly created clusters
+
+### Non-Goals
+
+- perform automated migration between using IMDSv1 and IMDSv2 for existing infrastructure
+
+## Proposal
+
+ - Add new options into install-config for Othe CP installation program which will be passed as metadata service settings during virtual machines creation
+ - Extend [AWSMachineProviderConfig](https://github.com/openshift/api/blob/master/machine/v1beta1/types_awsprovider.go#L12) by adding metadata service parameters
+ - Pass metadata service parameters to AWS during new machines creation
+
+### User Stories
+
+- As an OpenShift cluster administrator, I would like to enforce IMDSv2 usage during an OCP cluster deployment procedure.
+- As an OpenShift cluster administrator, I would like to enforce IMDSv2 usage for newly created machines in my cluster via MachineSets configuration.
+- As an OpenShift cluster administrator, I would like to change IMDS interaction settings for newly created machines in my cluster.
+
+### API Extensions
+
+#### AWSMachineProviderConfig changes
+
+MachineAPI design mostly replicates parameters which [AWS API](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceMetadataOptionsRequest.html) consumes.
+
+```go
+type AWSMachineProviderConfig struct {
+    // Existing fields will not be modified
+    ...
+	MetadataServiceOptions *MetadataServiceOptions `json:"metadataServiceOptions,omitempty"`
+}
+```
+
+```go
+
+type InstanceMetadataHttpTokens string
+
+const (
+    // InstanceMetadataHttpTokensRequired enforces sending of a signed token header with any instance metadata retrieval requests
+    InstanceMetadataHttpTokensRequired = "required"
+    // InstanceMetadataHttpTokensOptional allows IMDSv1 usage along with IMDSv2
+    InstanceMetadataHttpTokensOptional = "optional"
+)
+
+// MetadataServiceOptions defines the options available to a user when configuring
+// Instance Metadata Service (IMDS) Options.
+type MetadataServiceOptions struct {
+    // HttpTokens is the state of token usage for instance metadata requests from this host.
+    // Effectively enforces IMDSv2 if set to required.
+    // Default: optional
+    // +kubebuilder:validation:Enum=required;optional
+    // +optional
+    HttpTokens InstanceMetadataHttpTokens `json:"httpTokens,omitempty"`
+    // HttpPutResponseHopLimit is the desired HTTP PUT response hop limit for instance metadata requests.
+    // The larger the number, the further instance metadata requests can travel.
+    // Constraints: 1-64
+    // Default: 1
+    // +kubebuilder:validation:Minimum=1
+    // +kubebuilder:validation:Maximum=64
+    // +optional
+    HttpPutResponseHopLimit *uint8 `json:"httpPutResponseHopLimit,omitempty"`
```

Comment:
This doesn't need to be a pointer, and for conventions sake, I'd probably make this just an `int32`

Reply:
I would prefer to check for `nil` value, rather than for 0, which might be set there intentionally. With pointer, we have strong signal that value was not set and can use default value.

In regards of `int32`, why? Max value is 64 there, i do not se the reason to use `int32` here.

Reply:
> I would prefer to check for nil value, rather than for 0, which might be set there intentionally. With pointer, we have strong signal that value was not set and can use default value.

Are you not setting a default of 1? If you have a default there's no point in having a pointer, if you aren't setting a default and are leaving it to AWS to default, then a pointer makes more sense, but this needs to be clarified

> In regards of int32, why?

API conventions dictate normally to use int32. I don't think uint8 gains us very much, and, typically the utils we have for dereferencing pointers etc are all based on int32, so it may just be easier